linkml/generators/pydanticgen/array.py

```diff
@@ -0,0 +1,396 @@
+
+
+class ListOfListsArray(ArrayRangeGenerator):
+    """
+    Represent arrays as lists of lists!
+
+    TODO: Move all validation of values (eg. anywhere we raise a ValueError) to the ArrayExpression
+    dataclass and out of the generator class
+    """
+
+    REPR = ArrayRepresentation.LIST
+
+    @staticmethod
+    def _list_of_lists(dimensions: int, dtype: str) -> str:
+        return ("List[" * dimensions) + dtype + ("]" * dimensions)
+
+    @staticmethod
+    def _labeled_dimension(dimension: DimensionExpression, dtype: str) -> SlotResult:
+        # TODO: Preserve label representation in some readable way! doing the MVP now of using conlist
+        if dimension.exact_cardinality and (dimension.minimum_cardinality or dimension.maximum_cardinality):
+            raise ValueError("Can only specify EITHER exact_cardinality OR minimum/maximum cardinality")
+        elif dimension.exact_cardinality:
+            dmin = dimension.exact_cardinality
+            dmax = dimension.exact_cardinality
+        elif dimension.minimum_cardinality or dimension.maximum_cardinality:
+            dmin = dimension.minimum_cardinality
+            dmax = dimension.maximum_cardinality
+        else:
+            # TODO: handle labels for labeled but unshaped arrays
+            return SlotResult(annotation="List[" + dtype + "]")
+
+        items = []
+        if int(PYDANTIC_VERSION[0]) >= 2:
+            if dmin is not None:
+                items.append(f"min_length={dmin}")
+            if dmax is not None:
+                items.append(f"max_length={dmax}")
+        else:
+            if dmin is not None:
+                items.append(f"min_items={dmin}")
+            if dmax is not None:
+                items.append(f"max_items={dmax}")
+        items.append(f"item_type={dtype}")
+        items = ", ".join(items)
+        annotation = f"conlist({items})"
+
+        return SlotResult(annotation=annotation, imports=_ConListImports)
+
+    def any_shape(self, array: Optional[ArrayExpression] = None, with_inner_union: bool = False) -> SlotResult:
+        """
+        An AnyShaped array
+
+        Args:
+            array (:class:`.ArrayExpression`): The array expression (not used)
+            with_inner_union (bool): If ``True`` , the innermost type is a ``Union`` of the ``AnyShapeArray`` class
+                and ``dtype`` (default: ``False`` )
+
+        """
+        if self.dtype in ("Any", "AnyType"):
+            annotation = "AnyShapeArray"
+        else:
+            annotation = f"AnyShapeArray[{self.dtype}]"
+
+        if with_inner_union:
+            annotation = f"Union[{annotation}, {self.dtype}]"
+        return SlotResult(annotation=annotation, injected_classes=_AnyShapeArrayInjects, imports=_AnyShapeArrayImports)
+
+    def anonymous_shape(self, array: ArrayExpression) -> SlotResult:
+        if array.exact_number_dimensions:
+            return SlotResult(annotation=self._list_of_lists(array.exact_number_dimensions, self.dtype))
```

Comment:
```suggestion
        if array.exact_number_dimensions or (
            array.minimum_number_dimensions and array.maximum_number_dimensions and 
            array.minimum_number_dimensions == array.maximum_number_dimensions
        ):
            return SlotResult(annotation=self._list_of_lists(array.exact_number_dimensions, self.dtype))
```

Since the case min==max should be the same as if exact is specified, it might be nice to combine these conditions and avoid going into the below logic. This would also avoid creating a Union of annotations of length 1.

Reply:
ah, good catch